# MT2_Mm-CRISPRi

### "230829_ChIPseq_workflow_multi.groovy": 

  bpipe workflow for ChIP-seq analyses

### "230829_totalRNAseq_STARmap.groovy": 

  bpipe workflow for total RNA-seq analyses

### "230829_main1_sgRNAdesign_and_prediction.Rmd"

  R notebook for MT2_Mm sgRNA design, on- and off-targetting prediction analyses

### "230829_main2_dCas9_ChIP_analyses.Rmd"

  R notebook for dCas9/HA ChIP-seq analyses: peak filtering, annotation, overlap with repeats, and deeptool heatmap, etc

### "230829_main3_CRISPRi_total_RNAseq_analyses.Rmd"

  R notebook for RNA-seq QC, differential gene/repeat expression analyses

### "230829_main4_DEGs_relations_to_dCas9_ChIP.Rmd"

  R notebook for analyzing distances between DEGs and targeted repeats

### "230829_main5_TE_chimeric_analyses.Rmd"

  R notebook for LTR-chimeric analyses

### "230829_main6_Obox_analyses.Rmd"

  R notebook for minor ZGA, major ZGA, maternal gne, and Obox-related analyses

### "utils.R"

  R scritp for commonly used R functions
  

